### DISCUSSION TITLE
How do  I remove `IOE` effect from my static api?

### DISCUSSION THREAD

**Discussion by newhoggy** [haskell-effectful/effectful#276] [Q&A] [answered]
I have the following implementation that requires the `IOE` constraint:

```
sendAws :: forall a r. ()
  => HasCallStack
  => AwsRequest a
  => r <: Aws
  => r <: DataLog AwsLogEntry
  => r <: Error AwsError
  => r <: IOE
  => Typeable (AwsResponse a)
  => Typeable a
  => a
  -> Eff r (AwsResponse a)
sendAws req = GHC.withFrozenCallStack $ do
  envAws0 <- askAwsEnv

  envAws1 <- withEffToIO (ConcUnlift Persistent Unlimited) $ \effToIO ->
    pure $ envAws0
      { AWS.logger = \logLevel builder ->
          effToIO $ dataLog $ AwsLogEntry GHC.callStack logLevel builder
      }

  liftIO (runResourceT $ AWS.sendEither envAws1 req)
    & onLeftM throwError
```

How can I remove it?

I would like to retain the ability to have the implementation of my function execute in `Eff r` and not `IO`.

See https://github.com/haskell-works/effectful-zoo/pull/4/files#r1852196188

**Comment by arybczak** [comment, marked as answer]
You can't :thinking: 

Statically dispatched effects are supposed to be (for the lack of a better word) primitive, i.e. only use `IO` and data associated with your `StaticRep` instance. I think this needs more emphasis in the documentation.

If you find yourself in a situation of combining multiple effects in core functions of such effect, you actually want dynamic dispatch with all of its perks.

**Reply by newhoggy** [reply]
Thanks!
